samples/nrf9160/nrf_cloud_rest_agps/src/main.c

```diff
@@ -0,0 +1,604 @@
+	};
+	struct nrf_cloud_rest_context rest_ctx = {
+		.connect_socket = -1,
+		.keep_alive = true,
+		.timeout_ms = 30000,
+		.rx_buf = rx_buf,
+		.rx_buf_len = sizeof(rx_buf),
+		.fragment_size = 0
+	};
+	struct lte_lc_cells_info net_info = {
+		.ncells_count = 0,
+		.neighbor_cells = NULL
+	};
+	struct nrf_cloud_rest_agps_request agps_req = {
+		.type = NRF_CLOUD_REST_AGPS_REQ_CUSTOM,
+		.agps_req = &agps_request,
+		.net_info = &net_info
+	};
+	struct nrf_cloud_rest_agps_result agps_result = {
+		.buf_sz = 0,
+		.buf = NULL
+	};
+	int err;
+
+retry:
+	err = init();
+	if (err) {
+		LOG_ERR("Initialization failed");
+		goto cleanup;
+	}
+
+	err = connect_to_network();
+	if (err) {
+		goto cleanup;
+	}
+
+	err = get_modem_info(&modem_info);
+	if (err) {
+		goto cleanup;
+	}
+	current_cell.mcc = modem_info.network.mcc.value;
+	current_cell.mnc = modem_info.network.mnc.value;
+
+	agps_print_enable(false);
+
+	err = gps_start(gps_dev, &gps_cfg);
+	if (!err) {
+		LOG_INF("GPS STARTED!");
+	}
+
+	(void)do_jitp();
+
+agps_wait:
+	LOG_INF("Waiting for AGPS request event..");
+	k_sem_take(&do_agps_request_sem, K_FOREVER);
+
+	err = modem_jwt_generate(&jwt);
+	if (err) {
+		LOG_ERR("Failed to generate JWT, err %d", err);
+		goto cleanup;
+	}
+	LOG_DBG("JWT:\n%s", log_strdup(jwt.jwt_buf));
+	rest_ctx.auth = jwt.jwt_buf;
+
+	LOG_INF("\n********************* A-GPS API *********************");
+	/* When requesting a potentially large result, set fragment size to a
+	 * small value (1) and do not provide a result buffer.
+	 * This will allow you to obtain the necessary size while limiting
+	 * data transfer.
+	 */
+	net_info.current_cell = current_cell; /* update with latest cell tower info */
+	rest_ctx.fragment_size = 1;
+	agps_sz = nrf_cloud_rest_agps_data_get(&rest_ctx, &agps_req, NULL);
+	if (agps_sz < 0) {
+		LOG_ERR("Failed to get A-GPS data: %d", agps_sz);
+		goto agps_wait; /* one might want to retry the download instead */
+	} else if (agps_sz == 0) {
+		LOG_WRN("A-GPS data size is zero, skipping");
+		goto agps_wait; /* one might want to retry the download instead */
+	}
+
+	LOG_INF("Additional buffer required to download A-GPS data of %d bytes",
+		agps_sz);
+
+	agps_result.buf_sz = (uint32_t)agps_sz;
+	agps_result.buf = k_calloc(agps_result.buf_sz, 1);
+	if (!agps_result.buf) {
+		LOG_ERR("Failed to allocate %u bytes for A-GPS buffer", agps_result.buf_sz);
+		goto cleanup;
+	}
+	/* Use the default configured fragment size */
+	rest_ctx.fragment_size = 0;
+	err = nrf_cloud_rest_agps_data_get(&rest_ctx, &agps_req, &agps_result);
+	if (err) {
+		LOG_ERR("Failed to get A-GPS data: %d", err);
+	} else {
+		/* Send the A-GPS result buffer to the modem using
+		 * gps_process_agps_data() or nrf_cloud_agps_process()
+		 */
+		LOG_INF("Successfully received A-GPS data");
+		err = agps_cloud_data_process(agps_result.buf, agps_result.agps_sz);
+		if (err) {
+			LOG_ERR("Failed to process A-GPS data: %d", err);
+		}
+	}
+
+	k_free(agps_result.buf);
```

Comment:
can this chunk be combined into the code below `cleanup:`?
i think it should always cleanup and from there decide if it should reboot, sleep, or goto agps_wait?